json-rpc-lambda-client/src/main/java/com/amazonaws/blox/clients/JsonRpcLambdaClient.java

```diff
@@ -0,0 +1,193 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may
+ * not use this file except in compliance with the License. A copy of the
+ * License is located at
+ *
+ *     http://aws.amazon.com/apache2.0/
+ *
+ * or in the "LICENSE" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+package com.amazonaws.blox.clients;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.util.ByteBufferBackedInputStream;
+import com.googlecode.jsonrpc4j.IJsonRpcClient;
+import com.googlecode.jsonrpc4j.JsonRpcClient;
+import com.googlecode.jsonrpc4j.ProxyUtil;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import lombok.extern.slf4j.Slf4j;
+import software.amazon.awssdk.services.lambda.LambdaAsyncClient;
+import software.amazon.awssdk.services.lambda.model.InvokeRequest;
+import software.amazon.awssdk.services.lambda.model.InvokeResponse;
+
+/**
+ * A JSON-RPC client that translates calls into Lambda function invocations.
+ *
+ * <p>This class should not be used directly under normal circumstances. instead, use the {@link
+ * JsonRpcLambdaClient#newProxy(ObjectMapper, Class, LambdaAsyncClient, String) newProxy} factory
+ * method to create a dynamic client for a given interface.
+ */
+@Slf4j
+public class JsonRpcLambdaClient implements IJsonRpcClient {
+
+  private final LambdaAsyncClient lambda;
+  private final String functionName;
+  private final JsonRpcClient client;
+
+  private JsonRpcLambdaClient(LambdaAsyncClient lambda, String functionName) {
+    this(defaultObjectMapper(), lambda, functionName);
+  }
+
+  private JsonRpcLambdaClient(ObjectMapper mapper, LambdaAsyncClient lambda, String functionName) {
+    this.lambda = lambda;
+    this.functionName = functionName;
+    this.client = new JsonRpcClient(mapper);
+  }
+
+  private static ObjectMapper defaultObjectMapper() {
+    return new ObjectMapper();
+  }
+
+  /**
+   * Create a new dynamic JSON-RPC proxy for an interface that dispatches calls as Lambda function
+   * invocations.
+   *
+   * @param mapper the Jackson mapper to use for serializing/deserializing requests/responses
+   * @param serviceInterface the interface of the service to call
+   * @param lambdaClient a lambda client that can call the function that's running the service
+   *     implementation
+   * @param functionName the name of the labmda function that's running the service implementation
+   * @param <T>
+   * @return
+   */
+  public static <T> T newProxy(
```

Comment:
why not move this outside of this class so we don't have a mix of instantiated and static methods? it looks like some methods are calling the same methods on the client as well, it would help with readability imho to separate these out

Reply:
looks like renaming the client to jsonRPCclient or similar would make it clearer

Reply:
As we discussed, I'm trying to localize all of the unpleasant glue for this in one place, so that we don't have multiple classes that refer to the JSON-RPC implementation details spread through our application. I'll add the explanatory javadoc for that.